### ISSUE TITLE
Server documentation should use explicit buildProtocol()s rather than magic "protocol = MyProtocol" on factories

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4761]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#4761|
|Type|defect|
|Created|2010-12-06 11:41:25Z|
|Branch|https://github.com/twisted/twisted/tree/explicit-buildProtocol-4761|

Observation at the Twisted Lab in Boston suggests that buildProtocol as an important protocol/factory integration point is hard for new users to understand. This seems to be at least in part because the documentation tends to use the "protocol = MyProtocol" shortcut, which is basically magic and hides the existence buildProtocol from new developers.

All uses of factories in our documentation ought to use an explicit buildProtocol. The client and server howtos should document "protocol = MyProtocol" (but not use it otherwise).

<details><summary>Searchable metadata</summary>

```
trac-id__4761 4761
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_explicit_buildProtocol_4761 branches/explicit-buildProtocol-4761
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1291635685000000 1291635685000000
changetime__1302890846000000 1302890846000000
version__None None
owner__itamar itamar
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

+1.  I found that the explanation of buildProtocol as a method went over much more easily than the 'convenience' syntax of `protocol = MyProtocol`.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

The server howto doesn't even **mention** buildProtocol!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @itamarst|
|-|-|

Indeed not.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [31547]) Branching to 'explicit-buildProtocol-4761'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

I've decided to restrict this ticket to the server howto, since I'm running out of steam. I'll open a separate ticket for the client howto.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst removed owner|
|-|-|

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub set owner to @itamarst|
|-|-|

Thanks for this patch!

There is a simple error in the chat.py script. 'self.transport.write' is used instead of self.sendLine(). sendLine is much preferred because explicit line endings aren't required, and if the line endings change, what's transmitted is the same as what's received.

I dislike the part about how to limit the number of connections to the server. It's poor form for a network server to accept a connection, decide it didn't want to accept the connection, and drop the connection, instead buildProtocol fail and the connection should never be accepted. If you (or anyone else reading this review) agree this is the case, please file a ticket, as I think that's outside the scope of this enhancement.

There is a word in this document 'dynamicity', which is not in my spell checker, and perhaps this sentence could be phrased better. See: "Use Python's dynamicity to create open ...".

I have read through the changes to the prose and agree that it reads better and is clearer now, and I thoroughly appreciate the rewrite to get rid of the f = Factory(); f.protocol = MyProtocol idiom. Thank you!

I have manually checked the lore output, and it looks good, and there are no lint errors.

Please action the following:

 1. update the chat.py example and test it manually.
 1. file a ticket for the too many connections issue I raised, but only if you think it's worth changing (reference the number here if you did.
 1. Reword the dynamicity sentence.
 1. resubmit for review.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [31589]) Address review comments:

1. Replace transport.write with sendLine in chat example, and tested
manually. Also added sentence explaining sendLine where it is first
introduced.
2. Changed connection counter to not be used to reject connections.
3. Got rid of confusing sentence involving dynamicity; perhaps one day
we'll write full howto on state machines, but this can't be explained
one sentence.

References [#4761](https://github.com/twisted/twisted/issues/4761)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess set owner to @itamarst|
|-|-|

Thanks for working on this, itamar. A few comments:

 * I don't understand what the "which" is in "What this means is that the protocol responds to events as they arrive from the network, which arrives by calling functions on the protocol."
 * "the exists" ==> "that exists" in

```
+    instance. The factory is used to share state the exists beyond the
+    lifetime of any given connection. You will see why this object is
```


 * The example in "Simpler Protocol Creation" uses `listenTCP`, but `listenTCP` is only explained after its second use, in `listings/servers/chat.py`.
 * If the hope is that readers will put together the two bits of QOTD code from the Protocols section and have something runnable, some imports are missing:

```
from twisted.internet.protocol import Factory
from twisted.internet.endpoints import TCP4ServerEndpoint
from twisted.internet import reactor
```

 * The changes need a topfile.
 * I couldn't find the ticket for the client howto. Can you link it here?

Other than that, I think this is good to merge!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

Further work will be done when I figure out how to get open source IP release from Google (this may take a couple of weeks), unless someone else wants to take this up.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_13):
> Further work will be done when I figure out how to get open source IP release from Google (this may take a couple of weeks), unless someone else wants to take this up.

If / when you get that, please alert us if you will be contributing under Google's copyright or your own, so we know if we need to modify the copyright holder list in LICENSE.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> It's poor form for a network server to accept a connection, decide it didn't want to accept the connection, and drop the connection, instead buildProtocol fail and the connection should never be accepted.

This is wrong.  First, you cannot see where the connection is from until you accept it.  Second, if you don't accept it, the client is left waiting for a timeout before they know what's going on.  It's better to disconnect them immediately.  Best is to provide a _reason_ for disconnecting them, like _503 Service Unavailable_, but if this isn't possible actively accepting and then disconnecting still tells them the problem is with the server they are trying to connect, not some random intermediate router.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Thanks for the review jesstess.

> I don't understand what the "which"

Fixed

> "the exists" ==> "that exists"

Fixed

> The example in "Simpler Protocol Creation" uses `listenTCP`

I changed it to use an endpoint, like the earlier example.  I was tempted to make a similar change to `chat.py`, but that would have left the document with no examples of `listenTCP` which I ''guess' would be bad.

> If the hope is that readers will put together the two bits of QOTD

I added the missing imports.

> The changes need a topfile.

Looks like someone (itamar?) got to that first.

> I couldn't find the ticket for the client howto. Can you link it here?

It's [#5044](https://github.com/twisted/twisted/issues/5044)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [31668]) Address review feedback

refs [#4761](https://github.com/twisted/twisted/issues/4761)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [31670]) Merge explicit-buildProtocol-4761

Author: itamar
Reviewer: jesstess
Fixes: [#4761](https://github.com/twisted/twisted/issues/4761)

In the server howto, cover `Factory.buildProtocol` before introducing the
`Factory.protocol` shortcut.  This should make it more clear to new users
where protocols come from and what their relationship to factories is.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_15):
> > It's poor form for a network server to accept a connection, decide it didn't want to accept the connection, and drop the connection, instead buildProtocol fail and the connection should never be accepted.
> 
> This is wrong.  First, you cannot see where the connection is from until you accept it.

To clarify further, by the time `buildProtocol` is being called, the connection _has_ been `accept()`ed.  There's no way to avoid it.  [See the implementation here.](http://twistedmatrix.com/trac/browser/trunk/twisted/internet/tcp.py#L608)  The reason `buildProtocol` is allowed to return `None` is to avoid allocation work on an overloaded server on the _python_ side of things, not to change the behavior at the networking layer.
